### ISSUE TITLE
Hilt Android Plugin gradle configuration fails on kotlin 1.6.10 if app has transitive dependency on non-android library module

### ISSUE THREAD

**Issue by Wrywulf** [google/dagger#3125]
Kotlin 1.6.10
Dagger/Hilt: 2.40.5
Gradle: 7.3.1
AGP: 7.0.4

I have an app module `demo` with the hilt gradle plugin applied. It has a dependency chain this:
:features:mfc:demo > :base-ui > :base-utils

`:base-ui` applies "com.android.library" and "kotlin-kapt" plugins
`:base-utils` applies "java-library"  and "kotlin-kapt" plugins

So, `demo` has a transitive dependency on the base-utils module.

When compiling `demo` I get the following

```
* What went wrong:
Could not determine the dependencies of task ':features:mfc:demo:hiltJavaCompileDebug'.
> Could not resolve all task dependencies for configuration ':features:mfc:demo:hiltCompileOnlyDebug'.
   > Could not resolve project :base-utils.
     Required by:
         project :features:mfc:demo > project :base-ui
      > The consumer was configured to find a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'. However we cannot choose between the following variants of project :base-utils:
          - _classStructurekaptKotlin
          - _classStructurekaptTestKotlin
        All of them match the consumer attributes:
          - Variant '_classStructurekaptKotlin' capability root-project:base-utils:unspecified declares a runtime of a component:
              - Unmatched attribute:
                  - Doesn't say anything about com.android.build.api.attributes.BuildTypeAttr (required 'debug')
          - Variant '_classStructurekaptTestKotlin' capability root-project:base-utils:unspecified declares a runtime of a component:
              - Unmatched attribute:
                  - Doesn't say anything about com.android.build.api.attributes.BuildTypeAttr (required 'debug')
        The following variants were also considered but didn't match the requested attributes:
          - Variant 'apiElements' capability root-project:base-utils:unspecified:
              - Incompatible because this component declares an API of a component and the consumer needed a runtime of a component
              - Other compatible attribute:
                  - Doesn't say anything about com.android.build.api.attributes.BuildTypeAttr (required 'debug')
```

When I disable incremental compilation in gradle.properties the build succeeds:
```
kapt.incremental.apt=false
```
So I'm assuming that the bug is related to kapt somehow.

I don't understand why the `hiltCompileOnlyDebug` configuration would have the consumer requirement of `com.android.build.api.attributes.BuildTypeAttr` on java/kotlin only modules? 

Assembling `:base-ui` directly succeeds.

I did not have this problem before upgrading to kotlin 1.6.10. (was running 1.5.30 before).

**Comment by Wrywulf** [comment]
Is there anything I can help with in order to progress on this issue?

**Comment by danysantiago** [comment]
Sorry for the delay, but I have not been able to repro. We have a sample project with a similar setup as yours here: https://github.com/google/dagger/tree/master/javatests/artifacts/hilt-android/simpleKotlin

In it `app` depends on `android-library` and it depends on a Java/Kotlin only module called `deep-kotlin-lib`. Updating this project to the same versions you reported do not reproduce the issue.

Do you apply the Hilt plugin to `:base-ui` too?
Does `:base-utils` also has the `kotlin` plugin applied (aside from kapt and java ones)

Given a Kotlin versions change and a KAPT flag change reproduces the issue, have you filled a but to the Kotlin team at https://youtrack.jetbrains.com/issues?

To answer your question about `BuildTypeAttr`, the Hilt plugins requests for the attribute with a given value if present, but it is not required. We need to request it so if you have android-library modules with flavors, the right ones is used during aggregation.

Any chance you can reproduce in a project you can share?

**Comment by Wrywulf** [comment]
No problem, thanks for looking into it!

I have tried to reproduce by modifying the https://github.com/google/dagger/tree/master/javatests/artifacts/hilt-android/simpleKotlin by adding my own modules to it. But without any luck.

I'm guessing the problem I face is intrinsic to the setup in my codebase.

> Do you apply the Hilt plugin to :base-ui too?
No, it does not.

> Does :base-utils also has the kotlin plugin applied (aside from kapt and java ones)?
yes, it applies:
```
plugins {
    id("java-library")
    id("kotlin")
    id("kotlin-kapt")
}
```

I'll try to look more closely after a similar issue in the jetbrains issue tracker.

**Comment by GuilhE** [comment]
Same problem here using:
```
Kotlin 1.6.0
Dagger/Hilt: 2.40
Gradle: 7.3.3
AGP: 7.1.0-alpha05
```

**Comment by GuilhE** [comment]
Just updated to Bumblebee and using AGP 7.1.0 I've the same problem:

```
Could not determine the dependencies of task ':app:hiltJavaCompileDevDebug'.
> Could not resolve all task dependencies for configuration ':app:hiltCompileOnlyDevDebug'.
   > Could not resolve project :module-network.
     Required by:
         project :app
      > The consumer was configured to find a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug', attribute 'app' with value 'dev'. However we cannot choose between the following variants of project :module-network:
          - urlDevDebugRuntimeElements
          - urlProdDebugRuntimeElements
          - urlQaDebugRuntimeElements
        All of them match the consumer attributes:
          - Variant 'urlDevDebugRuntimeElements' capability android-kotlin-template:module-network:unspecified declares a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Unmatched attributes:
                  - Doesn't say anything about app (required 'dev')
                  - Provides attribute 'com.android.build.api.attributes.AgpVersionAttr' with value '7.1.0' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.api.attributes.ProductFlavor:module_network_dim' with value 'urlDev' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.gradle.internal.attributes.VariantAttr' with value 'urlDevDebug' but the consumer didn't ask for it
                  - Provides attribute 'module_network_dim' with value 'urlDev' but the consumer didn't ask for it
                  - Provides a library but the consumer didn't ask for it
                  - Provides attribute 'org.gradle.jvm.environment' with value 'android' but the consumer didn't ask for it
                  - Provides attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm' but the consumer didn't ask for it
          - Variant 'urlProdDebugRuntimeElements' capability android-kotlin-template:module-network:unspecified declares a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Unmatched attributes:
                  - Doesn't say anything about app (required 'dev')
                  - Provides attribute 'com.android.build.api.attributes.AgpVersionAttr' with value '7.1.0' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.api.attributes.ProductFlavor:module_network_dim' with value 'urlProd' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.gradle.internal.attributes.VariantAttr' with value 'urlProdDebug' but the consumer didn't ask for it
                  - Provides attribute 'module_network_dim' with value 'urlProd' but the consumer didn't ask for it
                  - Provides a library but the consumer didn't ask for it
                  - Provides attribute 'org.gradle.jvm.environment' with value 'android' but the consumer didn't ask for it
                  - Provides attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm' but the consumer didn't ask for it
          - Variant 'urlQaDebugRuntimeElements' capability android-kotlin-template:module-network:unspecified declares a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Unmatched attributes:
                  - Doesn't say anything about app (required 'dev')
                  - Provides attribute 'com.android.build.api.attributes.AgpVersionAttr' with value '7.1.0' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.api.attributes.ProductFlavor:module_network_dim' with value 'urlQa' but the consumer didn't ask for it
                  - Provides attribute 'com.android.build.gradle.internal.attributes.VariantAttr' with value 'urlQaDebug' but the consumer didn't ask for it
                  - Provides attribute 'module_network_dim' with value 'urlQa' but the consumer didn't ask for it
                  - Provides a library but the consumer didn't ask for it
                  - Provides attribute 'org.gradle.jvm.environment' with value 'android' but the consumer didn't ask for it
                  - Provides attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm' but the consumer didn't ask for it
        The following variants were also considered but didn't match the requested attributes:
          - Variant 'urlDevDebugApiElements' capability android-kotlin-template:module-network:unspecified declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Incompatible because this component declares an API of a component and the consumer needed a runtime of a component
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlDevReleaseApiElements' capability android-kotlin-template:module-network:unspecified:
              - Incompatible because this component declares an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlDevReleaseRuntimeElements' capability android-kotlin-template:module-network:unspecified declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlProdDebugApiElements' capability android-kotlin-template:module-network:unspecified declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Incompatible because this component declares an API of a component and the consumer needed a runtime of a component
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlProdReleaseApiElements' capability android-kotlin-template:module-network:unspecified:
              - Incompatible because this component declares an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlProdReleaseRuntimeElements' capability android-kotlin-template:module-network:unspecified declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlQaDebugApiElements' capability android-kotlin-template:module-network:unspecified declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
              - Incompatible because this component declares an API of a component and the consumer needed a runtime of a component
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlQaReleaseApiElements' capability android-kotlin-template:module-network:unspecified:
              - Incompatible because this component declares an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')
          - Variant 'urlQaReleaseRuntimeElements' capability android-kotlin-template:module-network:unspecified declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug'
              - Other compatible attribute:
                  - Doesn't say anything about app (required 'dev')

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

```

**:app**
```
flavorDimensions.add("app")
productFlavors {
    register("prod") {
        missingDimensionStrategy("module_network_dim", "urlProd")
    }

    register("qa") {
        ...
        missingDimensionStrategy("module_network_dim", "urlQa")
    }

    register("dev") {
        ...
        missingDimensionStrategy("module_network_dim", "urlDev")
    }
}
```
**:network**
```
flavorDimensions.add("module_network_dim")
productFlavors {
    register("urlProd") {
        ...
    }
    register("urlQa") {
        ...
    }
    register("urlDev") {
        ...
    }
}
```

**Comment by danysantiago** [comment]
@GuilhE it seems you are experiencing https://github.com/google/dagger/issues/3014 which should have been fixed for 7.1 based on the Google Issue: https://issuetracker.google.com/issues/205856789. Can you please report in the IssueTracker that you are still experiencing it.

**Comment by GuilhE** [comment]
> @GuilhE it seems you are experiencing #3014 which should have been fixed for 7.1 based on the Google Issue: https://issuetracker.google.com/issues/205856789. Can you please report in the IssueTracker that you are still experiencing it.

Done https://issuetracker.google.com/issues/205856789#comment9 😉

**Comment by personshelldon** [comment]
The same here on Android Studio Bumblebee:
```
Could not determine the dependencies of task ':app:hiltJavaCompileDefaultDebug'.
> Could not resolve all task dependencies for configuration ':app:hiltCompileOnlyDefaultDebug'.
   > Could not resolve project :translations.
     Required by:
         project :app
      > No matching variant of project :translations was found. The consumer was configured to find a runtime of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug', attribute 'operator' with value 'default' but:
          - Variant 'defaultDebugApiElements' capability Ministra:translations:unspecified declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug':
 ......
```

**Comment by Wrywulf** [comment]
Segue-waying back to my original problem.

What I've gathered is that the problem is directly related to the attributes of these variants on my java/kotlin (non-android) module also applying the `kotlin-kapt` plugin:

```
_classStructurekaptKotlin
_classStructurekaptTestKotlin
```
This is the `base-utils` module in the OP.

The interesting thing is that I get different declared capabilities for the above variants depending on if I enable or disable Kapt incremental processing:

```
./gradlew :base-utils:outgoingVariants --all -Pkapt.incremental.apt=true
--------------------------------------------------
Variant _classStructurekaptKotlin (l)
--------------------------------------------------
Capabilities
    - dart_gopass-app_android:base-utils:unspecified (default capability)
Attributes
    - org.gradle.usage = java-runtime

--------------------------------------------------
Variant _classStructurekaptTestKotlin (l)
--------------------------------------------------
Capabilities
    - dart_gopass-app_android:base-utils:unspecified (default capability)
Attributes
    - org.gradle.usage = java-runtime

```
And

```
./gradlew :base-utils:outgoingVariants --all -Pkapt.incremental.apt=false
--------------------------------------------------
Variant _classStructurekaptKotlin (l)
--------------------------------------------------
--------------------------------------------------
Variant _classStructurekaptTestKotlin (l)
--------------------------------------------------
```

The `java-runtime` attribute is what seems to be confusing gradle. I don't know why it is there in there first place and I definitely don't understand why it depends on the kapt incremental flag. Smells somewhat like a kapt regression to me unless hilt changed their variant matching recently.

The kotlin issue that I believe describes my issue best is this:
https://youtrack.jetbrains.com/issue/KT-30378

But i'm certainly no expert in this area, so I could be wrong.

@danysantiago do you think I should create a new issue for Kapt?

**Comment by Wrywulf** [comment]
FWIW I was able to workaround the problem by setting a dummy org.gradle.usage attribute on the unresolvable modules. Doing so bypasses the resolution completely since the configuration variants no longer have the runtime element:

```
configurations.configureEach {
    if (name == "_classStructurekaptTestKotlin" || name == "_classStructurekaptKotlin") {
        attributes {
            attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class, "dummy-usage"))
        }
    }
}
```

Don't know if there is a better workaround (deleing the attribute?), but the above works for me.

**Comment by danysantiago** [comment]
Hey - While fixing another issue we made a [change](https://github.com/google/dagger/commit/e3d446873e145f1e0b41f1fd08a67b4bca7cce0b) to Hilt's compile only configuration, specifically we no longer use AGP attributes which might have been causing some of the issues being described here.

It would be super cool if you could verify the issue is still present with the fix above by trying out a [snapshot version](https://oss.sonatype.org/content/repositories/snapshots/com/google/dagger/) of the plugin.

**Comment by personshelldon** [comment]
@WonderCsabo Hello! How I can test the SNAPSHOT version? I downloaded the SNAPSHOT gradle plugin, then I used this in project `build.gradle.kts`:
```
classpath(fileTree(mapOf("dir" to "$rootDir/dagger" ,"include" to listOf("hilt-android-gradle-plugin.jar"))))
```

But this does not work and on sync the  gradle says `The Hilt Android Gradle plugin is applied but no com.google.dagger:hilt-android dependency was found.`

How to import it into the project? Thanks.

**Comment by danysantiago** [comment]
To use the SNAPSHOT version, usually you add the repository to your `buildScript` repos:
`maven { url "https://oss.sonatype.org/content/repositories/snapshots/"}`, then depend on the snapshot version:
`com.google.dagger:hilt-android-gradle-plugin:HEAD-SNAPSHOT`, instructions might vary depending if your are using the newer plugin management declaration in `settings.gradle` or the `buildScript` closure, and also if your build files are in KTS.

**Comment by bcorso** [comment]
See https://dagger.dev/dev-guide/versions#head-snapshot-setup

**Comment by danysantiago** [comment]
Closing this, should be fixed in 2.42

**Comment by Wrywulf** [comment]
Also confirmed that 2.42 fixes it (apologize for the late confirmation)
